### PR TITLE
Computed config

### PR DESCRIPTION
**Pull request by davidlehn**
RFC: Update computed config API using a wrapper design.
There are README docs but not yet API docs.
Old API is not yet removed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+
+Normal code and the config API can be mixed. A useful helper is `require()`.
+This call lets you simplify ensuring a full object path exists before setting
+data. Objects in the path are created as needed.
+
+```js
+// create container object if needed
+c.require('accounts.admin');
+// the config is just a normal object
+config.accounts.admin.name = "Ima Admin";
```
Comment (mattcollier) [inline README.md]: I think this should be `bedrock.config....` for clarity.

File: README.md [hunk 2]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+
+```js
+// get the Config wrapper for the default bedrock.config
+var c = bedrock.util.config;
+// set static values
+c.set('server.port', 8443);
+c.set('server.domain', 'bedrock.dev');
+// set a computed value that uses values from the main config
+c.setComputed('server.host', () => {
+  return config.server.domain + ':' + config.server.port;
```
Comment (mattcollier) [inline README.md]: again, `bedrock.config...` or define config?

File: README.md [hunk 3]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+  // only add the port if it's not the well known default
+  if(config.server.port !== 443) {
+    return config.server.domain + ':' + config.server.port;
+  }
+  return config.server.domain;
+});
+```
+
+`setComputed()` can be verbose. a wrapper can be created using the standard
+`bind()` functionality. A helpper called `computer()` will do this for you.
```
Comment (mattcollier) [inline README.md]: typo: helpper

File: README.md [hunk 4]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+    return config.server.domain + ':' + config.server.port;
+  }
+  return config.server.domain;
+});
+```
+
+Copmuted values can also be created with string templates.
+
+```js
+cc('server.baseUri', 'https://${server.host}');
```
Comment (mattcollier) [inline README.md]: define `cc` again here so the example stands alone?

File: README.md [hunk 5]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
 insert their default configuration object using a key that either matches their
 module name or that matches their module name minus any `bedrock-` prefix. For
 example, the [bedrock-server][] module's specific configuration object can be
-found under `bedrock.config.server`. Modules may define whatever configuration
+found under `bedrock.config.server`. A `megacorp-feature` module would be found
+under `bedrock['megacorp-feature']`. Modules may define whatever configuration
```
Comment (dlongley) [inline README.md]: Let's do `mycompany-feature` instead.

File: README.md [hunk 6]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
 variables they want to using whatever format is appropriate for their own
 use.
 
+The `bedrock.util` module has helpers to setup configurations, and in
+particular, dynamic computed configurations. Computed values can help to
```
Comment (dlongley) [inline README.md]: dynamically

File: README.md [hunk 7]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+simplify dependency issues by allowing values to be computed at runtime from a
+function or string template. (Note there is a small cost with computed config
+values which could be important depending on the use case.)
+
+`bedrock.util.Config` creates a wrapper around a config object and optional
+options. A common setup could look like the following.
```
Comment (dlongley) [inline README.md]: `This wrapper exposes a new, helpful API that is detailed below. A common...`

File: README.md (hunk 1 above)
Comment (dlongley) [inline README.md]: Single quotes.

File: README.md [hunk 8]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+
+```js
+let c = bedrock.util.config;
+// set a config variable with a path
+// path components are created as needed
+c.set('server.port', 8443);
+// config is now {"server": {"port": 8443}}
+```
+
+Normal code and the config API can be mixed. A useful helper is `require()`.
```
Comment (dlongley) [inline README.md]: I actually think `require` is kind of a weird name. `ensure` is another name, but weird for other reasons. Require makes me think you're pulling something in, not making sure it exists.

Comment (mattcollier) [inline README.md]: `init`?

File: README.md [hunk 9]
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
...
+cc('server.host', () => {
+  // only add the port if it's not the well known default
+  if(config.server.port !== 443) {
+    return config.server.domain + ':' + config.server.port;
+  }
+  return config.server.domain;
+});
+```
+
+Copmuted values can also be created with string templates.
```
Comment (dlongley) [inline README.md]: Copmuted => Computed

Comment (dlongley) [inline README.md]: `with lodash-style string templates.`

**Comment by dlongley** [comment]
There are a few comments here to address, but this is looking really good! Can't wait to be using it everywhere and eliminate some headaches.

File: README.md (hunk 9 above)
Comment (mattcollier) [inline README.md]: Maybe a link to docs would be helpful.  If we use this syntax, it appears that the lodash website will redirect to the latest API version: `https://lodash.com/docs/#template`

File: README.md (hunk 8 above)
Comment (dlongley) [inline README.md]: `init` would be better too.

Comment (davidlehn) [inline README.md]: I think `init` implies you are fully initializing something vs just making sure it exists.  `init` could be done like set('foo.bar', {})

**Comment by davidlehn** [comment]
Forced update:
- Addressed comments.
- Changed `require()` to `setDefault(path, value)` or `setDefault({path, value, ...})`.
- Removed old API.
- Added README examples to tests.
- Fixed bugs.

**Pull request merged**